card/lib/card/migration/import.rb

```diff
@@ -0,0 +1,211 @@
+        # @option opts [String] remote Use a remote url. The remote url must
+        #   have been registered via 'add_remote'
+        # @option opts [Boolean] deep if true fetch all nested cards, too
+        # @option opts [Boolean] items_only if true fetch all nested cards but
+        #   not the card itself
+        def pull name, opts={}
+          MetaData.update do |meta_data|
+            url = opts[:remote] ? meta_data.url(opts.delete(:remote)) : nil
+            fetch_card_data(name, url, opts).each do |card_data|
+              saved_data = meta_data.add_card card_data
+              write_card_content saved_data[:key], card_data[:content]
+            end
+          end
+        end
+
+        # Save an url as remote deck to make it available for the pull method
+        def add_remote name, url
+          MetaData.update do |meta_data|
+            meta_data.remotes[name] = url
+          end
+        end
+
+        private
+
+        def card_data_for_merge all
+          MetaData.cards.map do |data|
+            next unless all || needs_update?(data)
+            key = data[:key] || data[:name].to_name.key
+            hash = {
+              'name' => data[:name],
+              'type' => data[:type],
+              'content' => File.read(content_path(key))
+            }
+            hash['codename'] = data[:codename] if data[:codename]
+            hash
+          end.compact
+        end
+
+        def write_card_content key, content
+          FileUtils.mkpath CARD_CONTENT_DIR unless Dir.exist? CARD_CONTENT_DIR
+          File.write content_path(key), content
+        end
+
+        def needs_update? data
+          !data[:pushed] ||
+            data[:pushed] < File.mtime(content_path(data[:name]))
+        end
+
+        # Returns an array of hashes with card attributes
+        def fetch_card_data name, url, opts
+          view, result_key =
+            if opts[:items_only]
+              ['export_items', nil]
+            elsif opts[:deep]
+              ['export', nil]
+            else
+              [nil, :card]
+            end
+
+          card_data =
+            if url
+              fetch_remote_data name, view, url
+            else
+              fetch_local_data name, view
+            end
+          result_key ? [card_data[result_key]] : card_data
+        end
+
+        def fetch_remote_data name, view, url
+          json_url = "#{url}/#{name}.json"
+          json_url += "?view=#{view}" if view
+          json = open(json_url).read
+          JSON.parse(json).deep_symbolize_keys
+        end
+
+        def fetch_local_data name, view
+          Card[name].format(format: :json).render(view || :content)
+        end
+
+        def content_path card_key
+          File.join CARD_CONTENT_DIR, card_key
+        end
+      end
+
+      # Handles the card attributes and remotes for the import
+      class MetaData
+        DEFAULT_PATH = Card::Migration.data_path('cards.yml').freeze
```

Comment:
I _think_ this is right.  I'm getting myself in a little tangle about whether this should be stored in the database vs in a file.  Because, does it really describe the state of the database or the filesystem?  It already stores some migration data in the db, right?